src/surface_tracker/location.py

```diff
@@ -0,0 +1,474 @@
+            surface._registered_markers_by_uid_undistorted
+        )
+
+        # Return None for an invalid surface definition
+        if len(registered_marker_uids) == 0:
+            logger.debug(f"Surface does not have any registered markers that define it")
+            return None
+
+        # Get the marker UIDs for the visible markers
+        visible_markers_by_uid = {m.uid: m for m in markers}
+        if len(visible_markers_by_uid) < len(markers):
+            raise ValueError(f"Detected markers must be unique")
+
+        # Only keep the visible marker UIDs for markers that define the surface
+        visible_markers_by_uid = {
+            uid: m
+            for uid, m in visible_markers_by_uid.items()
+            if uid in registered_marker_uids
+        }
+
+        # If the surface is defined by 1 marker, but it's missing - return
+        if len(registered_marker_uids) == 1 and len(visible_markers_by_uid) == 0:
+            return None
+        # If the surface is defined by 2 or more markers, but there are fewer than 2 markers visible - return
+        elif len(registered_marker_uids) >= 2 and len(visible_markers_by_uid) < 2:
+            return None
+
+        # Get the set of marker UIDs that are both visible and registered
+        matching_marker_uids = set(visible_markers_by_uid.keys())
+
+        marker_vertices_order = CornerId.all_corners()
+
+        visible_vertices_distorted = np.array(
+            [
+                visible_markers_by_uid[uid]._vertices_in_image_space(order=marker_vertices_order)
+                for uid in matching_marker_uids
+            ]
+        )
+        visible_vertices_distorted.shape = (-1, 2)
+
+        visible_vertices_undistorted = camera_model.undistort_points_on_image_plane(
+            visible_vertices_distorted
+        )
+        visible_vertices_undistorted.shape = (-1, 2)
+
+        registered_vertices_distorted = np.array(
+            [
+                registered_markers_by_uid_distorted[uid]._vertices_in_surface_space(order=marker_vertices_order)
+                for uid in matching_marker_uids
+            ]
+        )
+        registered_vertices_distorted.shape = (-1, 2)
+
+        registered_vertices_undistorted = np.array(
+            [
+                registered_markers_by_uid_undistorted[uid]._vertices_in_surface_space(order=marker_vertices_order)
+                for uid in matching_marker_uids
+            ]
+        )
+        registered_vertices_undistorted.shape = (-1, 2)
+
+        transform_matrix_from_image_to_surface_distorted, transform_matrix_from_surface_to_image_distorted = _find_homographies(
+            registered_vertices_distorted, visible_vertices_distorted
+        )
+
+        transform_matrix_from_image_to_surface_undistorted, transform_matrix_from_surface_to_image_undistorted = _find_homographies(
+            registered_vertices_undistorted, visible_vertices_undistorted
+        )
+
+        return _SurfaceLocation_v2(
```

Comment:
Similar to the other comment: Why not make this a classmethod and return
```suggestion
        return cls(
```
?

Reply:
In the case of `SurfaceLocation`, the `SurfaceLocation` class is a facade for concrete implementations.

`SurfaceLocation._create_location_from_markers` must return a concrete instance that implements the `SurfaceLocation` interface. The exact implementation that is returned here might change in the future (e.g. `_SurfaceLocation_v3`, etc.) and the only guarantee is that the returned instance implements the common interface.

Also, since the facade API implies calling the method on `SurfaceLocation`, making it a `classmethod` won't work, since `cls` in this case will be `SurfaceLocation`, which is an abstract class. The user would be required to explicitly call `_SurfaceLocation_v2._create_location_from_markers` which defeats the purpose of the facade in the first place.

Please let me know if this makes sense or not.

Reply:
Thank you for the detailed explanation. In the context of the abstract interface, this makes indeed a lot of sense!